src/Aspire.Hosting/Codespaces/CodespacesUrlRewriter.cs

```diff
@@ -0,0 +1,69 @@
+using Microsoft.Extensions.Logging;
+
+namespace Aspire.Hosting.Codespaces;
+
+internal class CodespacesUrlRewriter(ILogger<CodespacesUrlRewriter> logger, IConfiguration configuration, ResourceNotificationService resourceNotificationService) : BackgroundService
+{
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        if (!configuration.GetValue<bool>("CODESPACES", false))
+        {
+            logger.LogTrace("Not running in Codespaces, skipping URL rewriting.");
+            return;
+        }
+
+        var gitHubCodespacesPortForwardingDomain = configuration.GetValue<string>("GITHUB_CODESPACES_PORT_FORWARDING_DOMAIN") ?? throw new DistributedApplicationException("Codespaces was detected but GITHUB_CODESPACES_PORT_FORWARDING_DOMAIN environment missing.");
+        var codespaceName = configuration.GetValue<string>("CODESPACE_NAME") ?? throw new DistributedApplicationException("Codespaces was detected but CODESPACE_NAME environment missing.");
+
+        do
+        {
+            var resourceEvents = resourceNotificationService.WatchAsync(stoppingToken);
+
+            await foreach (var resourceEvent in resourceEvents.ConfigureAwait(false))
+            {
+                Dictionary<UrlSnapshot, UrlSnapshot>? remappedUrls = null;
+
+                foreach (var originalUrlSnapshot in resourceEvent.Snapshot.Urls)
+                {
+                    var uri = new Uri(originalUrlSnapshot.Url);
+
+                    if (!originalUrlSnapshot.IsInternal && (uri.Scheme == "http" || uri.Scheme == "https") && uri.Host == "localhost")
+                    {
+                        if (remappedUrls is null)
+                        {
+                            remappedUrls = new();
+                        }
+
+                        var newUrlSnapshot = originalUrlSnapshot with
+                        {
+                            Url = $"{uri.Scheme}://{codespaceName}-{uri.Port}.{gitHubCodespacesPortForwardingDomain}{uri.AbsolutePath}"
+                        };
+
+                        remappedUrls.Add(originalUrlSnapshot, newUrlSnapshot);
+                    }
+                }
+
+                if (remappedUrls is not null)
+                {
+                    var transformedUrls = from originalUrl in resourceEvent.Snapshot.Urls
+                                          select remappedUrls.TryGetValue(originalUrl, out var remappedUrl) ? remappedUrl : originalUrl;
+
+                    await resourceNotificationService.PublishUpdateAsync(resourceEvent.Resource, resourceEvent.ResourceId, s => s with
+                    {
+                        Urls = transformedUrls.ToImmutableArray()
+                    }).ConfigureAwait(false);
+                }
+            }
+
+            // Short delay if we crash just to avoid spinning CPU.
+            await Task.Delay(5000, stoppingToken).ConfigureAwait(false);
+        } while (!stoppingToken.IsCancellationRequested);
```

Comment:
What are you doing here? If `WatchAsync` errors then won't it throw an error right out of this method? Also, there should be a log message if there was an error and you need to recover.

Add a test that covers this situation to verify it works.